File: packages/rust/slop-ai/src/transport/unix.rs
//! Unix domain socket transport using NDJSON (newline-delimited JSON).
//!
//! ```no_run
//! use slop_ai::SlopServer;
//! use slop_ai::transport::unix;
//!
//! #[tokio::main]
//! async fn main() {
//!     let slop = SlopServer::new("my-app", "My App");
//!     let handle = unix::listen(&slop, "/tmp/slop/my-app.sock").await.unwrap();
//!     handle.await.unwrap();
//! }
//! ```

use std::path::Path;
use std::sync::Arc;

use serde_json::Value;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::UnixListener;
use tokio::sync::mpsc;
use tokio::task::JoinHandle;

use crate::error::{Result, SlopError};
use crate::server::{Connection, SlopServer};

enum ConnMessage {
    Send(Value),
    Close,
}

struct ChannelConnection {
    tx: mpsc::UnboundedSender<ConnMessage>,
}

impl Connection for ChannelConnection {
    fn send(&self, message: &Value) -> Result<()> {
        self.tx
            .send(ConnMessage::Send(message.clone()))
            .map_err(|_| SlopError::Transport("connection closed".into()))
    }

    fn close(&self) -> Result<()> {
        let _ = self.tx.send(ConnMessage::Close);
        Ok(())
    }
}

/// Listen for SLOP consumers on a Unix domain socket.
///
/// Returns a `JoinHandle` that resolves when the listener shuts down.
pub async fn listen(slop: &SlopServer, socket_path: &str) -> Result<JoinHandle<()>> {
    // Clean up stale socket
    let _ = std::fs::remove_file(socket_path);
    if let Some(parent) = Path::new(socket_path).parent() {
        std::fs::create_dir_all(parent).map_err(|e| SlopError::Transport(e.to_string()))?;
    }

    let listener =
        UnixListener::bind(socket_path).map_err(|e| SlopError::Transport(e.to_string()))?;

    // Set restrictive permissions
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let _ = std::fs::set_permissions(socket_path, std::fs::Permissions::from_mode(0o600));
    }

    let slop = slop.clone();

    let handle = tokio::spawn(async move {
        while let Ok((stream, _)) = listener.accept().await {
            let slop = slop.clone();
            tokio::spawn(async move {
                let (reader, mut writer) = tokio::io::split(stream);
                let (tx, mut rx) = mpsc::unbounded_channel::<ConnMessage>();
                let conn: Arc<dyn Connection> = Arc::new(ChannelConnection { tx });

                // Spawn a writer task that drains the channel into the Unix socket
                tokio::spawn(async move {
                    while let Some(msg) = rx.recv().await {
                        match msg {
                            ConnMessage::Send(val) => {
                                let mut line = serde_json::to_string(&val).unwrap_or_default();
                                line.push('\n');
                                if writer.write_all(line.as_bytes()).await.is_err() {
                                    break;
                                }
                            }
                            ConnMessage::Close => {
                                let _ = writer.shutdown().await;
                                break;
                            }
                        }
                    }
                });

                slop.handle_connection(conn.clone());

                let mut lines = BufReader::new(reader).lines();
                while let Ok(Some(line)) = lines.next_line().await {
                    let line = line.trim().to_string();
                    if line.is_empty() {
                        continue;
                    }
                    if let Ok(msg) = serde_json::from_str::<Value>(&line) {
                        slop.handle_message(&conn, &msg);
                    }
                }

                slop.handle_disconnect(&conn);
            });
        }
    });

    Ok(handle)
}

/// Check whether `id` is a valid descriptor filename stem.
/// See spec/core/transport.md §Local discovery.
fn is_valid_descriptor_stem(id: &str) -> bool {
    let len = id.len();
    if len == 0 || len > 64 {
        return false;
    }
    let bytes = id.as_bytes();
    if !(bytes[0].is_ascii_lowercase() || bytes[0].is_ascii_digit()) {
        return false;
    }
    bytes
        .iter()
        .all(|&b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'.' || b == b'_' || b == b'-')
}

/// Write a discovery descriptor to `~/.slop/providers/`.
///
/// The providers directory is created with mode 0700 and the descriptor is
/// written atomically (write to a same-directory temp file, then rename) with
/// mode 0600. See spec/core/transport.md §Local discovery.
pub fn register_provider(id: &str, name: &str, socket_path: &str) -> Result<()> {
    if !is_valid_descriptor_stem(id) {
        return Err(SlopError::Transport(format!(
            "provider id {id:?} is not a valid descriptor filename stem",
        )));
    }
    let home = dirs_home()?;
    let providers_dir = Path::new(&home).join(".slop").join("providers");
    std::fs::create_dir_all(&providers_dir).map_err(|e| SlopError::Transport(e.to_string()))?;

    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        std::fs::set_permissions(&providers_dir, std::fs::Permissions::from_mode(0o700))
            .map_err(|e| SlopError::Transport(e.to_string()))?;
    }

    let descriptor = serde_json::json!({
        "id": id,
        "name": name,
        "slop_version": "0.1",
        "transport": {"type": "unix", "path": socket_path},
        "pid": std::process::id(),
        "capabilities": ["state", "patches", "affordances", "attention", "windowing", "async", "content_refs"]
    });

    let final_path = providers_dir.join(format!("{id}.json"));
    let tmp_path = providers_dir.join(format!("{id}.json.tmp.{}", std::process::id()));
    std::fs::write(&tmp_path, serde_json::to_string_pretty(&descriptor)?)
        .map_err(|e| SlopError::Transport(e.to_string()))?;

    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        if let Err(e) = std::fs::set_permissions(&tmp_path, std::fs::Permissions::from_mode(0o600))
        {
            let _ = std::fs::remove_file(&tmp_path);
            return Err(SlopError::Transport(e.to_string()));
        }
    }

    std::fs::rename(&tmp_path, &final_path).map_err(|e| {
        let _ = std::fs::remove_file(&tmp_path);
        SlopError::Transport(e.to_string())
    })?;
    Ok(())
}

/// Remove a discovery descriptor from `~/.slop/providers/`.
pub fn unregister_provider(id: &str) -> Result<()> {
    if !is_valid_descriptor_stem(id) {
        return Ok(());
    }
    let home = dirs_home()?;
    let path = Path::new(&home)
        .join(".slop")
        .join("providers")
        .join(format!("{id}.json"));
    let _ = std::fs::remove_file(path);
    Ok(())
}

fn dirs_home() -> Result<String> {
    std::env::var("HOME").map_err(|_| SlopError::Transport("HOME not set".into()))
}

File: packages/go/slop-ai/unix.go
package slop

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"path/filepath"
	"regexp"
	"sync"
)

// descriptorFilenameRe matches the stem of a valid provider descriptor filename.
// See spec/core/transport.md §Local discovery.
var descriptorFilenameRe = regexp.MustCompile(`^[a-z0-9][a-z0-9._-]{0,63}$`)

type ndjsonConn struct {
	conn net.Conn
	mu   sync.Mutex
}

func (c *ndjsonConn) Send(msg any) error {
	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}
	data = append(data, '\n')
	c.mu.Lock()
	defer c.mu.Unlock()
	_, err = c.conn.Write(data)
	return err
}

func (c *ndjsonConn) Close() error {
	return c.conn.Close()
}

// UnixOption configures ListenUnix behavior.
type UnixOption func(*unixOpts)

type unixOpts struct {
	discovery bool
}

// WithDiscovery enables writing a discovery descriptor to ~/.slop/providers/.
func WithDiscovery(v bool) UnixOption {
	return func(o *unixOpts) { o.discovery = v }
}

// ListenUnix listens for SLOP consumers on a Unix domain socket using NDJSON.
// It blocks until the context is cancelled.
func ListenUnix(ctx context.Context, s *Server, socketPath string, opts ...UnixOption) error {
	cfg := &unixOpts{}
	for _, o := range opts {
		o(cfg)
	}

	// Clean up stale socket
	_ = os.Remove(socketPath)
	if err := os.MkdirAll(filepath.Dir(socketPath), 0o700); err != nil {
		return err
	}

	listener, err := net.Listen("unix", socketPath)
	if err != nil {
		return err
	}
	defer listener.Close()
	defer os.Remove(socketPath)

	// Restrictive permissions
	_ = os.Chmod(socketPath, 0o600)

	if cfg.discovery {
		_ = RegisterProvider(s.id, s.name, socketPath)
		defer UnregisterProvider(s.id)
	}

	// Close listener when context is done
	go func() {
		<-ctx.Done()
		listener.Close()
	}()

	for {
		conn, err := listener.Accept()
		if err != nil {
			if ctx.Err() != nil {
				return nil // context cancelled
			}
			return err
		}
		go handleNdjsonConn(ctx, s, conn)
	}
}

func handleNdjsonConn(ctx context.Context, s *Server, rawConn net.Conn) {
	conn := &ndjsonConn{conn: rawConn}
	s.HandleConnection(conn)
	defer s.HandleDisconnect(conn)

	scanner := bufio.NewScanner(rawConn)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		var msg map[string]any
		if json.Unmarshal([]byte(line), &msg) == nil {
			s.HandleMessage(ctx, conn, msg)
		}
	}
}

// RegisterProvider writes a discovery descriptor to ~/.slop/providers/.
//
// The provider directory is created with mode 0700 and the descriptor is
// written atomically (write to a same-directory temp file, then rename) with
// mode 0600. See spec/core/transport.md §Local discovery.
func RegisterProvider(id, name, socketPath string) error {
	if !descriptorFilenameRe.MatchString(id) {
		return fmt.Errorf("slop: provider id %q is not a valid descriptor filename stem (must match %s)", id, descriptorFilenameRe)
	}
	dir := filepath.Join(homeDir(), ".slop", "providers")
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return err
	}
	if err := os.Chmod(dir, 0o700); err != nil {
		return err
	}
	descriptor := map[string]any{
		"id":           id,
		"name":         name,
		"slop_version": "0.1",
		"transport":    map[string]any{"type": "unix", "path": socketPath},
		"pid":          os.Getpid(),
		"capabilities": []string{"state", "patches", "affordances", "attention", "windowing", "async", "content_refs"},
	}
	data, err := json.MarshalIndent(descriptor, "", "  ")
	if err != nil {
		return err
	}
	finalPath := filepath.Join(dir, id+".json")
	tmpPath := filepath.Join(dir, fmt.Sprintf("%s.json.tmp.%d", id, os.Getpid()))
	if err := os.WriteFile(tmpPath, data, 0o600); err != nil {
		return err
	}
	if err := os.Chmod(tmpPath, 0o600); err != nil {
		_ = os.Remove(tmpPath)
		return err
	}
	if err := os.Rename(tmpPath, finalPath); err != nil {
		_ = os.Remove(tmpPath)
		return err
	}
	return nil
}

// UnregisterProvider removes a discovery descriptor from ~/.slop/providers/.
func UnregisterProvider(id string) {
	if !descriptorFilenameRe.MatchString(id) {
		return
	}
	path := filepath.Join(homeDir(), ".slop", "providers", id+".json")
	_ = os.Remove(path)
}

func homeDir() string {
	if h := os.Getenv("HOME"); h != "" {
		return h
	}
	return "/tmp"
}
